Completed

potravniy opened the pull request:
Буду благодарен за критический разбор и рекомендации по улучшению.
Также буду признателен за идеи о причинах некорректной (с моей точки зрения) работы вывода отладочного console.log() в функциях some() и every() (внутри комментариев в коде).

potravniy committed:
Completed

potravniy committed:
Reduce() has bee developed.

potravniy commented:
Thank you for your comments in advance.

vpotravnyy commented on index.html:
```diff
@@ -13,23 +13,44 @@
+				return result = result || predicate(item, index, arr);
+			}, false)
 		}
 
```
console.log ведет себя тут неверно потому что приоритет у операций `+` и `||` одинаковый
`a + b || c`    ==    `(a + b) || c`     !==     `a + (b || c)`

Хорошим моментом выполнения задания является то, что функция predicate не вызывается больше раз чем нужно для ответа. Если найдем true - больше не тратим время

potravniy replied:
Спасибо, я действительно упустил момент с приоритетом.

potravniy replied:
"Хорошим моментом выполнения задания является то, что функция predicate не вызывается больше раз чем нужно для ответа." - На самом деле вызывается. Отладочный console.log() это показывает... Как отсечь лишнее - я пока не разобрался.

vpotravnyy replied:
Отладочный консоль лог ее то и вызывает )

vpotravnyy commented on index_reduce.html:
```diff
@@ -0,0 +1,141 @@
+	<script type="text/javascript">
+		function reduce(array, callback, initialValue){
+			if (array.length === 0 && initialValue !== undefined) { return initialValue; }
+			var array = array.filter(function(item){return item !== undefined});
```
Это верный подход, только лучше не фильтровать входной массив а просто не вызывать callback для пустых элементов.
Во первых так шустрее - O(n) вместо O(2n)
Во вторых так и делает Array.protottype.reduce:

```
[1,,2].reduce(function(acc, it, index) {
  console.log(index);  // Разница тут
  return acc+it 
}, 0)
```

Конечно, это создает определенные проблемы для следующих проверок - но это также можно пофиксить за O(n)

potravniy replied:
Понятно.
Основной целью фильтрации на входе было удовлетворение требований по входным ограничениям от MDN:
"Если массив пустой и аргумент initialValue не указан, будет брошено исключение TypeError. Если массив состоит только из одного элемента (независимо от его положения в массиве) и аргумент initialValue не указан, или если аргумент initialValue указан, но массив пустой, то будет возвращено одно это значение, без вызова функции callback." 
Массив состоящий из одних undefined отлавливать конечно можно и в рекурсии.....

vpotravnyy commented on index_reduce.html:
```diff
@@ -0,0 +1,141 @@
+					return reduceInner(previousValue, ++index, callback);
+				};
+			}
+		}
```
Великолепное решение, код отлично читается!
Есть только одно замечание, и то уж слишком краевой случай. (см выше)

vpotravnyy commented:
Практически безукоризненно, 1 минорный баг с неоптимальностью, рефакторинг не требуется - код хороший
5+

potravniy committed:
Reduce() has been updated.

Массив теперь не копируется, проверка undefined перенесена внутрь
рекурсии, код читается хуже. Надеюсь работает быстрее.

potravniy commented:
Reduce() изменен в соответствии с рекомендациями.
Массив не копируется, проверка undefined перенесена внутрь рекурсии.
Читается хуже, надеюсь работает быстрее.

vpotravnyy commented:
Без фильтра - намного лучше - так должно работать корректно
Вообще решение для уровня "студент" - уже великолепное
Но вскрылась еще пара случаев:
1) Если `currentValue = undefined` это еще не значит что он - дырка (попробуй `var arr = [ , undefined, ,]`)
2) Если дырки в начале массива, а `initialValue` нет - то `initialValue = array[0]` не прокатит

Сверившись со спекой заливаю в репозиторий вариант, полностью аналогичный работе `Array.prototype.reduce`
